src/java.base/share/classes/jdk/internal/module/ModuleInfoWriter.java

```diff
@@ -82,86 +85,86 @@ private ModuleInfoWriter() { }
+                            .mapToInt(mm -> MODULE_MODS_TO_FLAGS.getOrDefault(mm, 0))
+                            .reduce(0, (x, y) -> (x | y)));
+
+                    md.rawVersion().ifPresent(vs -> mb.moduleVersion(vs));
+
+                    // requires
+                    for (ModuleDescriptor.Requires r : md.requires()) {
+                        int flags = r.modifiers().stream()
+                                .mapToInt(REQUIRES_MODS_TO_FLAGS::get)
+                                .reduce(0, (x, y) -> (x | y));
+                        mb.requires(ModuleRequireInfo.of(
+                                cp.moduleEntry(cp.utf8Entry(r.name())),
+                                flags,
+                                r.rawCompiledVersion().map(cp::utf8Entry).orElse(null)));
+                    }
+
+                    // exports
+                    for (ModuleDescriptor.Exports e : md.exports()) {
+                        int flags = e.modifiers().stream()
+                                .mapToInt(EXPORTS_MODS_TO_FLAGS::get)
+                                .reduce(0, (x, y) -> (x | y));
+                        var targets = e.targets().stream().map(mn -> cp.moduleEntry(cp.utf8Entry(mn)))
+                                .toArray(ModuleEntry[]::new);
+                        mb.exports(ModuleExportInfo.of(
+                                cp.packageEntry(cp.utf8Entry(e.source())),
+                                flags,
+                                targets));
+                    }
+
+                    // opens
+                    for (ModuleDescriptor.Opens opens : md.opens()) {
+                        int flags = opens.modifiers().stream()
+                                .mapToInt(OPENS_MODS_TO_FLAGS::get)
+                                .reduce(0, (x, y) -> (x | y));
+                        var targets = opens.targets().stream().map(mn -> cp.moduleEntry(cp.utf8Entry(mn)))
+                                .toArray(ModuleEntry[]::new);
+                        mb.opens(ModuleOpenInfo.of(
+                                cp.packageEntry(cp.utf8Entry(opens.source())),
+                                flags,
+                                targets));
+                    }
+
+                    // uses
+                    md.uses().stream().map(ClassDesc::of).forEach(mb::uses);
+
+                    // provides
+                    for (ModuleDescriptor.Provides p : md.provides()) {
+                        mb.provides(ClassDesc.of(p.service()),
+                                                 p.providers().stream()
+                                                         .map(ClassDesc::of)
+                                                         .toArray(ClassDesc[]::new));
+                    }
+                }));
+
+                // packages
+                var packages = md.packages().stream().sorted().map(PackageDesc::of).toList();
+                if (!packages.isEmpty()) {
```

Comment:
This check should be `if (md.packages().size() > exportedOrOpen)`

Reply:
An oddity in ModuleInfoWriter that I think dates from when a prototype version of this attribute contains the set of concealed packages. There's no need for ModuleInfoWriter to emit the ModulePackages attribute, Adam has chosen to have it emit it. It has implications for one test for ModuleDescriptor.read, and he's changed that.

Reply:
Ok.   We should move this out to the test library as you suggested since it's only used by the tests.  I will create a JBS issue for it.